vm/src/stdlib/socket.rs

```diff
@@ -0,0 +1,238 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+pub struct Socket {
+    af: AddressFamily,
+    sk: SocketKind,
+    tcp_listener: Option<TcpListener>,
+    tcp_stream: Option<TcpStream>,
+    udp_socket: Option<UdpSocket>,
+}
+
+impl Socket {
+    fn new(af: AddressFamily, sk: SocketKind) -> Socket {
+        Socket {
+            af: af,
+            sk: sk,
+            tcp_listener: None,
+            tcp_stream: None,
+            udp_socket: None,
+        }
+    }
+
+    fn get_tcp_stream(&self) -> Option<&TcpStream> {
+        match &self.tcp_stream {
+            Some(v) => Some(v),
+            None => None,
+        }
+    }
+}
+
+fn socket_new(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
+    arg_check!(
+        vm,
+        args,
+        required = [
+            (cls, None),
+            (family_int, Some(vm.ctx.int_type())),
+            (kind_int, Some(vm.ctx.int_type()))
+        ]
+    );
+
+    let family = AddressFamily::from_i32(objint::get_value(family_int).to_i32().unwrap());
+    let kind = SocketKind::from_i32(objint::get_value(kind_int).to_i32().unwrap());
+
+    let socket = Socket::new(family, kind);
+
+    Ok(PyObject::new(
+        PyObjectPayload::Socket { socket },
+        cls.clone(),
+    ))
+}
+
+fn socket_connect(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
+    arg_check!(
+        vm,
+        args,
+        required = [(zelf, None), (address, Some(vm.ctx.str_type()))]
+    );
+
+    let mut mut_obj = zelf.borrow_mut();
+
+    match mut_obj.payload {
+        PyObjectPayload::Socket { ref mut socket } => {
+            if let Ok(stream) = TcpStream::connect(objstr::get_value(&address)) {
+                socket.tcp_stream = Some(stream);
+                Ok(vm.get_none())
+            } else {
+                // TODO: Socket error
+                Err(vm.new_type_error("socket failed".to_string()))
+            }
+        }
+        _ => Err(vm.new_type_error("".to_string())),
+    }
+}
+
+fn socket_recv(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
+    arg_check!(
+        vm,
+        args,
+        required = [(zelf, None), (bufsize, Some(vm.ctx.int_type()))]
+    );
+    let mut mut_obj = zelf.borrow_mut();
+
+    match mut_obj.payload {
+        PyObjectPayload::Socket { ref mut socket } => match socket.af {
+            AddressFamily::AF_INET => match socket.sk {
+                SocketKind::SOCK_STREAM => {
+                    let mut buffer = Vec::new();
+                    socket
+                        .get_tcp_stream()
+                        .unwrap()
+                        .read_to_end(&mut buffer)
+                        .unwrap();
+                    Ok(PyObject::new(
+                        PyObjectPayload::Bytes { value: buffer },
+                        vm.ctx.bytes_type(),
+                    ))
+                }
+                _ => Err(vm.new_type_error("".to_string())),
```

Comment:
Instead of match you can use `if let`. Please run clippy on your code to get these kind of tips.